modules/sockets/sockets.cpp

```diff
@@ -20,228 +20,299 @@
-    server.sin_family = AF_INET;
-    // Change the port to network byte order, and put it into the structure
-    server.sin_port = htons(port);
-    
-    // Not, let's try to open a connection to the server
-    contr = connect(sock, (struct sockaddr*)&server, sizeof( server));
-    if (contr < 0) {
-            // If an error occured cancel
-            *err = 3;  //error while connecting
-            return -1;
-    }
-    // Everything went well, so return the socket
-    return sock;
-}
-
-// native socket_close(_socket);
-static cell AMX_NATIVE_CALL socket_close(AMX *amx, cell *params)  /* 2 param */
-{
-    int socket = params[1];
-    //PRINT_CONSOLE("Function: Close | Socket: %i\n", socket);
-    #ifdef _WIN32 // On windows, check whether the sockets are initialized correctly
-    closesocket(socket);
-    #else
-    // Close the socket (linux/unix styled systems)
-    close(socket);
-    #endif
-    return 0;
-}
-
-// native socket_change(_socket, _timeout=100000);
-// 1 sec =1000000 usec
-static cell AMX_NATIVE_CALL socket_change(AMX *amx, cell *params)  /* 2 param */
+#include <sh_string.h>
+
+// Initial size for recv() function's output buffer.
+//
+#define INITIAL_RECV_SIZE (8192)
+
+// Socket type for socket() function.
+//
+enum AmxxSocketType
 {
-    int socket = params[1];
-    unsigned int timeout = params[2];
-    //PRINT_CONSOLE("Function: Change | Socket: %i | Timeout: %i\n", socket, timeout);
-    // We need both a timeout structure and a fdset for our filedescriptor
-    fd_set rfds;
-    struct timeval tv;
-    // Fill in ...
-    FD_ZERO(&rfds);
-    FD_SET(socket, &rfds);
-    tv.tv_sec = 0;
-    tv.tv_usec = timeout;
-    // Now we "select", which will show us if new data is waiting in the socket's buffer
-    if (select(socket+1, &rfds, NULL, NULL, &tv) > 0) 
-       return 1; // Ok, new data, return it
-     else 
-       return 0; // No new data, return it
-}
-
-// native socket_recv(_socket, _data[], _length);
-static cell AMX_NATIVE_CALL socket_recv(AMX *amx, cell *params)  /* 2 param */
+	SOCKET_TCP = 1,
+	SOCKET_UDP
+};
+
+// socket_open(Hostname[], Port, Protocol = SOCKET_TCP, &errorNum)
+//
+static cell AMX_NATIVE_CALL socket_open(AMX * pAmx, cell * pParams)
 {
-    int socket = params[1];
-    int length = params[3];
-    int tmp = -1;
-    // First we dynamicly allocate a block of heap memory for recieving our data
-    char *tmpchar = new char[length];
-    if(tmpchar == NULL) return -1; // If we didn't got a block, we have to quit here to avoid sigsegv
-    // And set it all to 0, because the memory could contain old trash
-    memset(tmpchar, 0, length);
-    // Now we recieve
-    tmp = recv(socket, tmpchar, length-1, 0);
-	if (tmp == -1)
+	// Gets hostname, hostname's length and error address.
+	//
+	int Length = 0;
+	char * pName = MF_GetAmxString(pAmx, pParams[1], 0, &Length);
+	cell * pError = MF_GetAmxAddr(pAmx, pParams[4]);
+
+	// Hostname parameter is empty.
+	//
+	if (Length < 1)
```

Comment:
If name is empty after the socket has been created, just close the socket and return -1.
The socket must not stay initialized.